### PR TITLE
refactor(api-docs): rework ApiExplorerPage to use EntityListProvider

### LINKED ISSUE #4637: API Catalog: Add the ability to customize table columns and filters
Add the ability for users to specify custom table columns, and the ability to
filter by them, to the API catalog plugin.

## Feature Suggestion

Having the ability to add custom fields inside of an API descriptor file's
`metadata` opens up the possibility for displaying this data in the API catalog.
Naturally, we would have the ability to filter by this data.

## Possible Implementation

Backstage already defines a set of default columns to display and filter by in
the API catalog. In addition to these, we would allow users to define any
additional columns and/or filters. We could allow the user to specify a list of
column names, if the columns are part of the default set, and `TableColumn`s
that denote the additional, custom columns to display. Filters would behave in
the same way.

For example, suppose that I want to display all of the default columns, with the
exception of 'System', and add a column to display the business domain that an
API belongs to. I would declare it like so:

```ts 
const domainColumn = { title: 'Domain', field: 'entity.metadata.domain' };
const customColumns = [
  'Name',
  'Description',
  'Owner',
  'Lifecycle',
  'Type',
  domainColumn,
];
```

My original implementation was based off `v0.4.3` of the plugin and added
`columns` and `filters` props to the `ApiExplorerPage` component. On `v0.4.4`,
the plugin changed to use the new [composability system](https://backstage.io/docs/plugins/composability#extensions), so it doesn't seem like
using props is the way to go with routable extensions, which `ApiExplorerPage`
now is. Any pointers for making these customizations available and configurable
via the host app are greatly appreciated.

## Context

At the moment, we have a business need to display a pair of custom columns that
denote the domain and capabilities for each of the APIs that we're listing in
the catalog, as well as filter by these fields. I did some work in modifying the
plugin to let us do this, with the intention of contributing the changes to
Backstage. This work was originally meant to happen in #3560 (full disclosure -
@tradcliffe-expediagroup is a colleague of mine), but the `api-docs` plugin
declares its own columns and filters.

### LINKED ISSUE #5310: APIs explorer with pagination
## Feature Suggestion

Add pagination that is supported by `catalog-backend` to `api-docs` plugin.

## Possible Implementation
Extend `api-docs` components and `catalog-client` to support pagination.

## Context
At Zalando, we would like to add `api-docs` plugin to our system. The absence of pagination prevents us to introduce it since we are having >1k APIs. If there is no ongoing work or if it would be useful to other people, I would be happy to contribute to it.

### PR DESCRIPTION
**Pull request by kuangp**
Signed-off-by: Phil Kuang <pkuang@factset.com>

## Hey, I just made a Pull Request!
Rework `ApiExplorerPage` to ultilize `EntityListProvider` to provide a consistent UI with the `CatalogIndexPage` which now exposes support for starring entities, pagination, and customizing columns. 

Closes #4637

Screen Shot 2021-06-10 at 12 59 19 PM

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/api-docs/src/components/ApiExplorerTable/ApiExplorerTable.tsx
Comment (timbonicus) [inline plugins/api-docs/src/components/ApiExplorerTable/ApiExplorerTable.tsx:129]: This is losing functionality in that the `owner` and `lifecycle` filters are removed.

**Review by timbonicus — COMMENTED** [review]
Hey, this is a really awesome contribution! 😍 So cool to see a continuation of the `EntityListProvider` work.

The big concern here, I think, is losing the `lifecycle` and `owner` filters that exist today on the api-docs page. Of course these filters _also_ make sense on the CatalogPage, so I'd expect to create them in `catalog-react` alongside the other Picker components, so they could be used in both places. These could possibly register simple frontend filters for now, especially as **backend** `owner` filtering depends on #4499.

And - taking that mentally one step further - there's some interesting overlap with the _existing_ `OwnerPicker` component in the `scaffolder` plugin. Maybe not enough to sensibly consolidate these, but they should at least borrow from each other I would imagine.

To mentally go _even further_,  the ApiDocsPage _itself_ should become unnecessary once there's a UI for `EntityKindPicker` available on the CatalogPage. This makes sense as an interim step, though.

File: plugins/api-docs/src/components/ApiExplorerTable/ApiExplorerTable.tsx
Comment (kuangp) [reply plugins/api-docs/src/components/ApiExplorerTable/ApiExplorerTable.tsx:129]: Ah, good point, I'll look into implementing these

**Comment by timbonicus** [comment]
Noting too that I'm not 100% sure this closes #5310, since that mentions using the pagination exposed by `catalog-backed`. The catalog page, and therefore this PR as well, do client-side pagination rather than backend pagination.

**Comment by kuangp** [comment]
> Noting too that I'm not 100% sure this closes #5310, since that mentions using the pagination exposed by `catalog-backed`. The catalog page, and therefore this PR as well, do client-side pagination rather than backend pagination.

good point, I'll remove that issue from this PR

**Comment by kuangp** [comment]
@timbonicus updated this change with implementations for lifecycle and owner filters

**Review by timbonicus — APPROVED** [review]
Thanks Phil! This is looking really good, thanks for adding those missing filters. I hope you don't mind, I added one commit - primarily to add some spacing around the filters, and some minor language tweaks. I also added these new filters to the `CatalogPage` 😍 

One more minor comment -- and we'll need @backstage/maintainers for final approval. I think this is just about ready to go, though!

**Comment by kuangp** [comment]
> Thanks Phil! This is looking really good, thanks for adding those missing filters. I hope you don't mind, I added one commit - primarily to add some spacing around the filters, and some minor language tweaks. I also added these new filters to the `CatalogPage` 😍
> 
> One more minor comment -- and we'll need @backstage/maintainers for final approval. I think this is just about ready to go, though!

thanks for the updates!

File: plugins/catalog-react/src/components/EntityOwnerPicker/EntityOwnerPicker.tsx
Comment (kuangp) [reply plugins/catalog-react/src/components/EntityOwnerPicker/EntityOwnerPicker.tsx:43]: that definitely makes sense, let me know if we want to accommodate this now or if we're ok with shipping the implementation as is and enhancing later.

**Comment by kuangp** [comment]
@timbonicus really sorry but I rebased my branch and forced pushed without pulling down your commit and accidentally erased your changes. Do you still happen to have the commit locally to re-push onto this branch?

File: plugins/catalog-react/src/components/EntityOwnerPicker/EntityOwnerPicker.tsx
Comment (timbonicus) [reply plugins/catalog-react/src/components/EntityOwnerPicker/EntityOwnerPicker.tsx:43]: Brought this up with maintainers, would rather switch to using relations for this before merge 🙏

Comment (kuangp) [reply plugins/catalog-react/src/components/EntityOwnerPicker/EntityOwnerPicker.tsx]: @timbonicus updated the `EntityOwnerPicker` to utilize the `ownedBy` relation

File: plugins/api-docs/src/components/ApiExplorerPage/ApiExplorerPage.tsx
Comment (Rugvip) [inline plugins/api-docs/src/components/ApiExplorerPage/ApiExplorerPage.tsx:67]: Tbh I actually think it's best to keep the existing implementation in place here instead of re-using the button from the catalog plugin. It's not a log of code, and we end up with an individual switch and external route ref for the api docs plugin, which with this gets coupled together with the catalog plugin.

File: plugins/catalog/src/index.ts
Comment (Rugvip) [inline plugins/catalog/src/index.ts:21]: I'd vote for renaming this to `CatalogTableRow` as part of exporting it, to group things up a bit more

**Review by Rugvip — COMMENTED** [review]
Thank you, good stuff indeed! 👍

**Comment by kuangp** [comment]
@Rugvip updated per your review comments, thanks!

**Review by timbonicus — APPROVED** [review]
Thanks for the hard work on this, added tests, etc. It's really appreciated! 🙌

**Review by Rugvip — APPROVED** [review]
Thank you, let's :shipit:!

**Pull request merged**
